File: easyCDFT.py
# -*- coding: utf-8 -*-
"""
easyCDFT.py
Briefly describe the functionality and purpose of the file.

This is a Main function file!

This file is part of EasyCDFT.
EasyCDFT is a Python script that automate the use of Multiwfn

@author:
Kimariyb (kimariyb@163.com)

@license:
Licensed under the MIT License.
For details, see the LICENSE file.

@Data:
2023-09-13
"""
import configparser
import subprocess
import os

from datetime import datetime

# 获取当前文件被修改的最后一次时间
time_last = os.path.getmtime(os.path.abspath(__file__))
# 全局的静态变量
__version__ = "v1.0.0"
__developer__ = "Kimariyb, Ryan Hsiun"
__address__ = "XiaMen University, School of Electronic Science and Engineering"
__website__ = "https://github.com/kimariyb/easy-cdft"
__release__ = str(datetime.fromtimestamp(time_last).strftime("%b-%d-%Y"))


def welcome():
    """
    主页面信息
    """

    print(f"@Name: EasyCDFT")
    print(f"@Version: {__version__}, @Release date: {__release__}")
    print(f"@Developer: {__developer__}")
    print(f"@Address: {__address__}")
    print(f"@Website: {__website__}")

    # 获取当前日期日期和时间
    now_time = datetime.now().strftime("%b-%d-%Y, %H:%M:%S")
    # 输出版权信息和问候语
    print(f"(Copyright (C) 2023 Kimariyb. Currently timeline: {now_time})\n")


def read_config(config_path: str) -> configparser.ConfigParser():
    """
    读取当前文件夹下的 config.ini 文件

    Args:
        config_path(str): config.ini 文件的路径

    Returns:
        config(configparser.ConfigParser()): 创建 ini config 对象
    """
    # 获取 ini config 对象
    config = configparser.ConfigParser()
    # 设置注释前缀
    config.comment_prefixes = ';'
    # 添加一个虚拟的 section 头部
    config.read_string('[DEFAULT]\n' + open(config_path, encoding='utf-8').read())

    return config


def get_files_byType(input_type: str) -> list[str]:
    """
    得到符合 config 中 inputType 属性配置的文件类型的所有文件

    Args:
        input_type(str): 输入文件的类型

    Returns:
        results(list[str]): 返回符合要求的所有文件
    """
    # 得到当前文件夹路径
    current_path = os.getcwd()

    # 存储符合要求的文件
    results = []

    # 遍历当前文件夹下的所有文件
    for file_name in os.listdir(current_path):
        # 判断文件名的后缀是否与输入的类型匹配
        if file_name.endswith(input_type):
            # 如果匹配，则将文件名添加到结果列表中
            results.append(file_name)

    # 返回符合要求的所有文件
    return results


def process_file(file: str, multiwfn_path: str, commands: list[str]):
    """
    自动化执行 Multiwfn 程序

    Args:
        file(str): 需要处理的文件，可以为 fchk、gjf、mol 等文件
        multiwfn_path(str): Multiwfn 程序的路径
        commands(list[str]): 需要执行的命令

    Returns:
        None
    """
    print(f"Documents being processed: {file}")
    # 创建一个命令对象，执行 Multiwfn 程序
    cmd = [multiwfn_path, file]

    # 启动命令并获取标准输入和输出管道
    proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                            universal_newlines=True)

    # 向命令的标准输入写入指令
    for command in commands:
        proc.stdin.write(command + "\n")

    # 关闭标准输入管道
    proc.stdin.close()

    # 等待命令执行完成
    proc.wait()

    # 在每次执行完Multiwfn后会在当前文件夹下生成一个CDFT.txt文件
    # 将该文件名修改为 ${name}-CDFT.txt 文件，${name} 为一个变量，代表 file 文件的名字
    output_file = os.path.splitext(file)[0] + "-CDFT.txt"
    # 重命名 CDFT.txt 文件
    os.rename("CDFT.txt", output_file)
    print("Renamed CDFT.txt to", output_file)


def main():
    # 显示 hello 页面
    welcome()
    # config.ini 文件路径，必须在运行程序的当前路径下，名字为 config.ini
    config_path = os.path.join(os.getcwd(), 'config.ini')
    # 读取 config.ini 文件
    config = read_config(config_path)
    # 读取 inputType 和 multiwfnPath 两个配置
    input_type = str(config.get('DEFAULT', 'inputType')).strip('\"')
    multiwfn_path = str(config.get('DEFAULT', 'multiwfnPath')).strip('\"')
    # 同时打印 inputType 和 multiwfnPath 两个配置
    print(f"inputType: {input_type}")
    print(f"multiwfnPath: {multiwfn_path}\n")
    # 根据 inputType 所配置的文件类型，得到当前文件下所有符合 inputType 类型的文件
    files = get_files_byType(input_type)
    print(f"Files with inputType {input_type}:")
    # 打印每一个文件的名字
    for file in files:
        print(file)
    # 读取 mission 配置
    mission = int(config.get('DEFAULT', 'mission'))
    # 读取 calcLevel 配置
    calc_level = str(config.get('DEFAULT', 'calcLevel')).strip('\"')
    # 读取 ChargeSpin1、2、3 配置
    chargeSpin1 = str(config.get('DEFAULT', 'chargeSpin1')).strip('\"')
    chargeSpin2 = str(config.get('DEFAULT', 'chargeSpin2')).strip('\"')
    chargeSpin3 = str(config.get('DEFAULT', 'chargeSpin3')).strip('\"')

    # 根据 config.ini 设置储存命令
    if mission == 0:
        # 如果 mission 为 0 则使用计算指数命令
        commands = ["22", "1", calc_level, chargeSpin1, chargeSpin2, chargeSpin3, "y", "2", "q"]
    elif mission == 1:
        print("Warning: This feature is not yet developed")
        return
    elif mission == 2:
        print("Warning: This feature is not yet developed")
        return
    else:
        # 如果没有符合的命令，则抛出异常
        print("Error: Invalid mission value")
        return
    # 批量执行 Multiwfn
    for file in files:
        process_file(file, multiwfn_path, commands)

    print()
    print("The mission has been successfully completed!")


if __name__ == '__main__':
    main()

File: easyCDFT.go
package main

import (
	"fmt"
	"github.com/go-ini/ini"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

/*
EasyCDFT 是厦门大学 Kimariyb 开发的一款全自动批处理使用 Multiwfn 计算概念密度泛函理论 CDFT 各种量的 Go 语言程序

实现思路：
	1. 首先获取当前文件夹下的所有文件，找到符合用户在 config.ini 中设定的文件类型
	2. 批量用 Multiwfn 执行选中的文件
	3. 读取文件后，使用流实现目的

@Name: EasyCDFT
@Author: Kimariyb
@Institution: XiaMen University
@Data: 2023-09-12
*/

/*
展示开始界面，包括版权，作者，版本等信息
*/
func show() {
	// 获取当前文件的绝对路径
	filePath, err := filepath.Abs(os.Args[0])
	if err != nil {
		fmt.Println("获取文件路径失败:", err)
		return
	}

	// 获取最后修改时间的时间戳
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		fmt.Println("获取文件信息失败:", err)
		return
	}

	modTime := fileInfo.ModTime()
	timestamp := modTime.Format("2006-Jan-02")

	fmt.Println("@Name: EasyCDFT")
	fmt.Println("@Version: v1.0.0, @Release date:", timestamp)
	fmt.Println("@Developer: Kimariyb, Ryan Hsiun")
	fmt.Println("@Address: XiaMen University, School of Electronic Science and Engineering")
	fmt.Println("@Website: https://github.com/kimariyb/easy-cdft")

	// 获取当前日期和时间
	now := time.Now().Format("Jan-02-2006, 15:04:05")

	// 输出版权信息和问候语
	fmt.Printf("(Copyright (C) 2023 Kimariyb. Currently timeline: %s)\n", now)
	fmt.Println()
}

/*
读取当前文件夹下的 config.ini 文件
*/
func readIniFile(filePath string) (*ini.File, error) {
	cfg, err := ini.Load(filePath)
	if err != nil {
		return nil, err
	}
	return cfg, nil
}

/*
获取 config 文件的路径
*/
func getConfigFilePath() string {
	currentPath, _ := os.Getwd()
	configFileName := "config.ini"
	configFilePath := filepath.Join(currentPath, configFileName)
	return configFilePath
}

/*
得到符合 config 中 inputType 属性配置的文件类型的所有文件
*/
func getFilesByType(inputType string) ([]string, error) {
	currentPath, _ := os.Getwd()
	files, err := ioutil.ReadDir(currentPath)
	if err != nil {
		return nil, err
	}

	var result []string
	for _, file := range files {
		if !file.IsDir() && filepath.Ext(file.Name()) == "."+inputType {
			result = append(result, file.Name())
		}
	}

	return result, nil
}

func processFile(file string, multiwfnPath string, commandLines []string) {
	fmt.Println("Documents being processed:", file)

	// 创建一个命令对象，执行 Multiwfn 程序
	cmd := exec.Command(multiwfnPath, file)

	// 获取标准输入管道
	stdin, err := cmd.StdinPipe()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// 启动命令
	err = cmd.Start()
	if err != nil {
		fmt.Println("Error starting process:", err)
		return
	}

	// 向命令的标准输入写入指令
	for _, line := range commandLines {
		_, err := io.WriteString(stdin, line+"\n")
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
	}

	// 关闭标准输入管道
	stdin.Close()

	// 等待命令执行完成
	err = cmd.Wait()

	// 在每次执行完 Multiwfn 后会在当前文件夹下生成一个 CDFT.txt 文件
	// 将该文件名修改为 ${name}-CDFT.txt 文件，${name} 为一个变量，代表 file 文件的名字
	// 生成 CDFT.txt 文件名
	outputFile := strings.TrimSuffix(file, filepath.Ext(file)) + "-CDFT.txt"
	// 重命名 CDFT.txt 文件
	err = os.Rename("CDFT.txt", outputFile)
	if err != nil {
		fmt.Println("Error renaming file:", err)
		return
	}

	fmt.Println("Renamed CDFT.txt to", outputFile)
}

func main() {
	// 展示 hello 页面
	show()
	//config.ini 文件路径，必须在运行程序的当前路径下，名字为 config.ini
	configFilePath := getConfigFilePath()
	// 读取 config 文件
	cfg, err := readIniFile(configFilePath)
	if err != nil {
		fmt.Printf("Failed to read INI file: %v\n", err)
		return
	}
	// 读取 ini 文件中的 section
	section, err := cfg.GetSection("")
	if err != nil {
		fmt.Printf("Failed to get section: %v\n", err)
		return
	}
	// 读取 inputType 和 multiwfnPath 两个配置
	inputType := section.Key("inputType").String()
	multiwfnPath := section.Key("multiwfnPath").String()
	// 同时打印 inputType 和 multiwfnPath
	fmt.Printf("inputType: %s\n", inputType)
	fmt.Printf("multiwfnPath: %s\n", multiwfnPath)
	// 根据 inputType 所配置的文件类型，得到当前文件下所有符合 inputType 类型的文件
	files, err := getFilesByType(inputType)
	if err != nil {
		fmt.Printf("Failed to get files: %v\n", err)
		return
	}
	fmt.Println("Files with inputType", inputType+":")
	for _, file := range files {
		fmt.Println(file)
	}

	//读取 mission 配置
	mission := section.Key("mission").String()
	// 读取 calcLevel 配置
	calcLevel := section.Key("calcLevel").String()
	// 读取 chargeSpin1、2、3 配置
	chargeSpin1 := section.Key("chargeSpin1").String()
	chargeSpin2 := section.Key("chargeSpin2").String()
	chargeSpin3 := section.Key("chargeSpin3").String()
	// 定义需要实现的流，创建一个切片用于存放命令
	var commandLines []string
	// 根据 config.ini 设置储存命令
	if mission == "0" {
		// 如果 mission 为 0 则使用计算指数命令
		commandLines = append(commandLines, "22", "1", calcLevel, chargeSpin1, chargeSpin2, chargeSpin3, "y", "2", "q")
	} else if mission == "1" {
		// 如果 mission 为 1 则使用计算福井函数命令
		fmt.Println("Warning: This feature is not yet developed")
		return
	} else if mission == "2" {
		// 如果 mission 为 2 则使用
		fmt.Println("Warning: This feature is not yet developed")
		return
	} else {
		// 如果没有符合的命令，则抛出异常
		fmt.Println("Error: Invalid mission value")
		// 返回错误或中止程序
		return
	}

	// 批量执行 Multiwfn
	for _, file := range files {
		processFile(file, multiwfnPath, commandLines)
	}

	fmt.Println()
	fmt.Println("The mission has been successfully completed!")

}
